src/api/c/memory.cpp

```diff
@@ -1,26 +1,99 @@
 /*******************************************************
- * Copyright (c) 2014, ArrayFire
+ * Copyright (c) 2019, ArrayFire
  * All rights reserved.
  *
  * This file is distributed under 3-clause BSD license.
  * The complete license agreement can be obtained at:
  * http://arrayfire.com/licenses/BSD-3-Clause
  ********************************************************/
 
+#include <memoryapi.hpp>
+
 #include <Array.hpp>
 #include <backend.hpp>
 #include <common/err_common.hpp>
 #include <handle.hpp>
-#include <memory.hpp>
 #include <platform.hpp>
 #include <af/backend.h>
 #include <af/device.h>
 #include <af/dim4.hpp>
+#include <af/memory.h>
 #include <af/version.h>
-#include <cstring>
 
 using namespace detail;
 
+BufferInfo &getBufferInfo(const af_buffer_info handle) {
+    return *(BufferInfo *)handle;
+}
+
+af_buffer_info getHandle(BufferInfo &pair) {
+    BufferInfo *handle;
+    handle = &pair;
+    return (af_buffer_info)handle;
+}
+
+af_err af_create_buffer_info(af_buffer_info *handle, void *ptr,
+                             af_event event) {
+    try {
+        BufferInfo *pair = new BufferInfo({ptr, event});
+        *handle          = getHandle(*pair);
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_release_buffer_info(af_buffer_info handle) {
+    try {
+        /// NB: deleting a memory event pair does NOT free the associated memory
+        /// and does NOT delete the associated event.
+        delete (BufferInfo *)handle;
```

Comment:
Restarting the converstion from [here](https://github.com/arrayfire/arrayfire/pull/2461#discussion_r311190544):

> This plan sounds reasonable. I'll add these functions to the C++ API as well.
> 
> In terms of implementation: unlocking the event is simple, but in terms of freeing the pointer, should that be accomplished with `memFree` directly? Can't see another way.

Custom memory managers should be using these functions. they are responsible for releasing the memory using nativeAlloc and nativeFree.

Reply:
Ended up using `af_free_device` [here](https://github.com/arrayfire/arrayfire/pull/2461/commits/48d454edeba1230dcc0a391d4d1665e892edb9c0#diff-fdfb1b65d24289c1211389e3d0c78ec8R57) which will call the memory manager anyways, as you said.